src/networking/serialization/mod.rs

```diff
@@ -0,0 +1,31 @@
+
+pub mod bits;
+pub mod rawend;
+
+use bincode;
+use crate::networking::payloads::payload::Ipv8Payload;
+use bincode::ErrorKind;
+use serde::de::Deserialize;
+use crate::networking::serialization::rawend::RawEnd;
+use serde::ser::Serialize;
+
+/// Deserializes a stream of bytes into an ipv8 payload. Which payload is inferred by the type of T which is generic.
+/// T has to be deserializable and implement the Ipv8Payload trait.
+pub fn deserialize<T>(buffer: &[u8]) -> Result<T, Box<ErrorKind>>
+  where for<'de> T: Deserialize<'de> + Ipv8Payload {
+  let mut cur = &buffer[..];
+
+  let mut msg: T = bincode::deserialize_from(&mut cur)?;
+  msg.set_rawend(RawEnd(cur.to_owned()));
+
+  Ok(msg)
+}
+
+/// simple wrapper function to serialize to bincode. TODO: how will we handle serialization to other standards like json easily?
```

Comment:
Unless I'm mistaken `serde_json` _should_ handle that without much trouble.

Reply:
well yes it would but right now its doing bincode::deserialize which would ofc only work for bincode. some generic thing would be better

Reply:
IPv8 doesn't use JSON, therefore we don't have to support it. Solving the task at hand with as little lines of code as possible is always cleaner than doing future engineering. You can always change it later if you want, as long as you're not in production.